tpu-client/src/nonblocking/tpu_client.rs

```diff
@@ -447,6 +453,37 @@ where
         })
     }
 
+    pub fn rpc_client(&self) -> &RpcClient {
+        &self.rpc_client
+    }
+
+    pub async fn shutdown(&mut self) {
+        self.exit.store(true, Ordering::Relaxed);
+        self.leader_tpu_service.join().await;
+    }
+}
+
+impl<P, M, C> TpuClient<P, M, C>
+where
+    P: ConnectionPool<NewConnectionConfig = C> + Send + Sync + 'static,
+    M: ConnectionManager<ConnectionPool = P, NewConnectionConfig = C> + Send + Sync + 'static,
+    C: Send + Sync + 'static,
+{
+    /// Send a transaction to the current and upcoming leader TPUs according to fanout size
+    /// Returns a detached running task with the last error if all sends fail
+    pub async fn try_send_wire_transaction_detached(
+        &self,
+        wire_transaction: Vec<u8>,
+    ) -> JoinHandle<TransportResult<()>> {
+        let leaders = self
+            .leader_tpu_service
+            .leader_tpu_sockets(self.fanout_slots);
+        let connection_cache = self.connection_cache.clone();
+        task::spawn(async move {
```

Comment:
In the past we have run into issues like spawning too many tasks into the runtime as such the the runtime has trouble handle all the tasks and some tasks like PING (keep-alive) for QUIC was delayed and causing connection timeout. That led us to adding throttling in the quic-client. Have we done enough stress test and see if we have similar concerns? Maybe using bench-tps with tpu-client and --thread-batch-sleep-ms set to 0 and with many threads for example (20). Also, do we have bench-tps results of using tpu-client with and without the fix?

Reply:
Gotcha, where can I find the throttling? I don't see it in `quic-client/src/nonblocking/quic_client.rs`. Right now, this is sending transactions with the TPU client in one multi-threaded tokio runtime, and waiting 10ms in between each send.

In 1.14, sending a transaction with the TPU client is almost instantaneous, so the goal of this PR is to fix the regression for users. The test was doing 100 self-transfers with the TPU client. Without this PR, it takes 18s-28s on devnet, and with this PR it takes 2.5-3.5s. When I profiled a little closer, I noticed that sending takes 100ms-250ms on devnet or testnet per transaction.

I haven't tested this change in bench-tps just yet, since that wasn't the focus, but I would love to try it there once this lands!

Reply:
The throttling code using a semaphore: https://github.com/solana-labs/solana/blob/master/quic-client/src/quic_client.rs#L160. The bench-tps test I referred to is the local cluster bench-tps test which can be obtained either manually via net script or through buildkite: https://buildkite.com/solana-labs/solana-local-cluster/builds/

Reply:
Great thanks for this. Just so I understand though, are you suggesting that I use this new function in bench-tps and see if it breaks the tokio runtime with --thread-batch-sleep-ms=0 and ~20 threads?

Unless there's something I'm missing, let's do that in follow-up work and first fix the regression that users are experiencing. This code generates a maximum of 100 transactions per second, which is very safely within the bounds of the 2000 tasks defined in the semaphore.

Reply:
Can you point to me the logic we do 100 txns per second? In my test with local cluster using tpu-client with bench-tps I was seeing much higher TPS.